src/go/rpk/pkg/config/v21_1_4.go

```diff
@@ -0,0 +1,276 @@
+		Redpanda:	*redpandaConf,
+		Rpk:		c.Rpk.toGeneric(),
+	}
+	return conf, nil
+}
+
+func (rc v2114RedpandaConfig) toGeneric() (*RedpandaConfig, error) {
+	rpcServer := rc.RPCServer.toGeneric()
+
+	var advRpcApi *SocketAddress
+	if rc.AdvertisedRPCAPI != nil {
+		advRpcApi = rc.AdvertisedRPCAPI.toGeneric()
+	}
+
+	var kafkaApi []NamedSocketAddress
+	if rc.KafkaApi != nil {
+		var err error
+		kafkaApi, err = v2114ParsePolymorphicSocketAddress(rc.KafkaApi)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var advKafkaApi []NamedSocketAddress
+	if rc.AdvertisedKafkaApi != nil {
+		var err error
+		advKafkaApi, err = v2114ParsePolymorphicSocketAddress(rc.AdvertisedKafkaApi)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kafkaApiTls := rc.KafkaApiTLS.toGeneric()
+
+	adminApi := rc.AdminApi.toGeneric()
+
+	var seeds []SeedServer
+	if rc.SeedServers != nil {
+		for _, s := range rc.SeedServers {
+			seeds = append(seeds, s.toGeneric())
+		}
+	}
+
+	return &RedpandaConfig{
+		Directory:		rc.Directory,
+		RPCServer:		*rpcServer,
+		AdvertisedRPCAPI:	advRpcApi,
+		KafkaApi:		kafkaApi,
+		AdvertisedKafkaApi:	advKafkaApi,
+		KafkaApiTLS:		kafkaApiTls,
+		AdminApi:		*adminApi,
+		Id:			rc.Id,
+		SeedServers:		seeds,
+		DeveloperMode:		rc.DeveloperMode,
+	}, nil
+}
+
+func (rpk v2114RpkConfig) toGeneric() RpkConfig {
+	return RpkConfig{
+		TLS:				rpk.TLS.toGeneric(),
+		AdditionalStartFlags:		rpk.AdditionalStartFlags,
+		EnableUsageStats:		rpk.EnableUsageStats,
+		TuneNetwork:			rpk.TuneNetwork,
+		TuneDiskScheduler:		rpk.TuneDiskScheduler,
+		TuneNomerges:			rpk.TuneNomerges,
+		TuneDiskWriteCache:		rpk.TuneDiskWriteCache,
+		TuneDiskIrq:			rpk.TuneDiskIrq,
+		TuneFstrim:			rpk.TuneFstrim,
+		TuneCpu:			rpk.TuneCpu,
+		TuneAioEvents:			rpk.TuneAioEvents,
+		TuneClocksource:		rpk.TuneClocksource,
+		TuneSwappiness:			rpk.TuneSwappiness,
+		TuneTransparentHugePages:	rpk.TuneTransparentHugePages,
+		EnableMemoryLocking:		rpk.EnableMemoryLocking,
+		TuneCoredump:			rpk.TuneCoredump,
+		CoredumpDir:			rpk.CoredumpDir,
+		WellKnownIo:			rpk.WellKnownIo,
+		Overprovisioned:		rpk.Overprovisioned,
+		SMP:				rpk.SMP,
+	}
+}
+
+func (sa *v2114SocketAddress) toGeneric() *SocketAddress {
+	return &SocketAddress{
+		Address:	sa.Address,
+		Port:		sa.Port,
+	}
+}
+
+func (sa *v2114NamedSocketAddress) toGeneric() *NamedSocketAddress {
+	addr := sa.v2114SocketAddress.toGeneric()
+	return &NamedSocketAddress{
+		SocketAddress:	*addr,
+		Name:		sa.Name,
+	}
+}
+
+func (stls v2114ServerTLS) toGeneric() ServerTLS {
+	return ServerTLS{
+		KeyFile:	stls.KeyFile,
+		CertFile:	stls.CertFile,
+		TruststoreFile:	stls.TruststoreFile,
+		Enabled:	stls.Enabled,
+	}
+}
+
+func (tls v2114TLS) toGeneric() TLS {
+	return TLS{
+		KeyFile:	tls.KeyFile,
+		CertFile:	tls.CertFile,
+		TruststoreFile:	tls.TruststoreFile,
+	}
+}
+
+func (s v2114SeedServer) toGeneric() SeedServer {
+	host := s.Host.toGeneric()
+	return SeedServer{Host: *host}
+}
+
+func v2114ParsePolymorphicSocketAddress(
```

Comment:
For functions like this, it should be possible to use [mapstructure DecodeHooks](https://pkg.go.dev/github.com/mitchellh/mapstructure#DecoderConfig). The goal would be to `mapstructure.Decode(theWholeConfig)`.

Reply:
I think there are two aspects:

### Transformations

E.g.,  the map/list for advertised - hopefully we can have a single canonical hierarchical representation that maps to the config struct, they should have a duality.

Then there could be other schema, such as a flattened map - possibly [RecursiveStructToMapHookFunc](https://pkg.go.dev/github.com/mitchellh/mapstructure#RecursiveStructToMapHookFunc) does that. Or older config schema if they ever need to be written. But reading old config should be possible with other transformations; either by inspection of the type, or possibly by inspection of the version.

### Validations

E.g., a list of NamedSocketAddress each has a name and those names are unique.

### Composition

It may not make sense for transformations and validations to be intertwined, but one advantage if they are, is that it might be possible to emit a more specific error.